pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,149 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cdi "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ClientFactory creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+type ClientFactory interface {
+	// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+	// It also returns the namespace of the kubeconfig's current context.
+	GetClient(kubeconfig []byte) (client.Client, string, error)
+}
+
+// ClientFactoryFunc is a function that implements ClientFactory.
+type ClientFactoryFunc func(kubeconfig []byte) (client.Client, string, error)
+
+// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+// It also returns the namespace of the kubeconfig's current context.
+func (f ClientFactoryFunc) GetClient(kubeconfig []byte) (client.Client, string, error) {
+	return f(kubeconfig)
+}
+
+// DataVolumeManager manages Kubevirt DataVolume operations.
+type DataVolumeManager interface {
+	// CreateOrUpdateDataVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+	CreateOrUpdateDataVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error
+	// GetDataVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+	// not found object error.
+	GetDataVolume(ctx context.Context, kubeconfig []byte, name string) (*cdi.DataVolume, error)
+	// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+	ListDataVolumes(ctx context.Context, kubeconfig []byte, listOpts ...client.ListOption) (*cdi.DataVolumeList, error)
+	// DeleteDataVolume delete the DataVolume based on the passed name and namespace.
+	DeleteDataVolume(ctx context.Context, kubeconfig []byte, name string) error
+}
+
+type defaultDataVolumeManager struct {
+	client ClientFactory
+	logger logr.Logger
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFactory) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+		logger: log.Log.WithName("kubevirt-data-volume-manager"),
+	}, nil
+}
+
+// CreateOrUpdateDataVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateOrUpdateDataVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error {
+	c, namespace, err := d.client.GetClient(kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, "could not create kubevirt client")
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, c, dataVolume, func() error {
+		dataVolume.Namespace = namespace
```

Comment:
I am afraid this won't work. You *must* set the namespace outside `CreateOrUpdate` so that this function finds the correct object if it exists. Also you *must* mutate the object spec (but not name or namespace) inside the `mutateFn` for reasons already explained. Did you try following the recommendations from my previous review?
> To achieve this, you could pass the name and spec of the data volume as separate args, set the name / namespace before calling CreateOrUpdate, and set the spec inside the mutateFn.